File: pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // get card number (between 13 and 20 digits)
    long card;
    do
    {
        card = get_long("Number: ");
    }
    while (!(card > 0));

    // get length of card
    int length;
    int div;
    for (int i = 20; i > 0; i--)
    {
        div = card / (long) pow(10, i - 1);
        if (div > 0)
        {
            length = i;
            break;
        }
    }

    // get checksum
    int checksum = 0;
    // - rule 1: Multiply every other digit by 2, starting with the number’s second-to-last digit,
    // and then add those products’ digits together.
    for (int i = 1; i < 20; i += 2)
    {
        long topBoundry = card % (long) pow(10, i + 1);
        long bottomBoundry = card % (long) pow(10, i);
        int num = (topBoundry - bottomBoundry) / pow(10, i);

        num *= 2;

        if (num >= 10)
        {
            int ones = num % 10;
            int tens = num / 10;
            checksum += (tens + ones);
        }
        else
        {
            checksum += (num);
        }
    }

    // - rule 2: Add the sum to the sum of the digits that weren’t multiplied by 2.
    for (int i = 0; i < 20; i += 2)
    {
        long topBoundry = card % (long) pow(10, i + 1);
        long bottomBoundry = card % (long) pow(10, i);
        int num = (topBoundry - bottomBoundry) / pow(10, i);

        checksum += num;
    }

    // - rule 3: If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is
    // congruent to 0), the number is valid!
    int amex_start = card / (long) pow(10, 15 - 2);
    int mc_start = card / (long) pow(10, 16 - 2);
    int visa_start_13 = card / (long) pow(10, 13 - 1);
    int visa_start_16 = card / (long) pow(10, 16 - 1);

    if (checksum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (length == 15 && (amex_start == 34 || amex_start == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (mc_start == 51 || mc_start == 52 || mc_start == 53 ||
                              mc_start == 54 || mc_start == 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && (visa_start_13 == 4 || visa_start_16 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/sentimental-credit/credit.py
from cs50 import get_int, get_string

# get card number
card = None
while not isinstance(card, int) or card < 0:
    try:
        card = get_int("Number: ")
    except:
        pass

# Convert card to str datatype
card = str(card)
length = len(card)

# Perform checksum
checksum = 0

# - rule 1: Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
for i in range(1, length, 2):
    x = int(card[-i-1]) * 2

    if x >= 10:  # exception case where the multiplication is > 10, and only the digits are added
        tens = x // 10
        ones = x % 10
        checksum += (tens + ones)
    else:
        checksum += x

# - rule 2: Add the sum to the sum of the digits that weren’t multiplied by 2.
for i in range(0, length, 2):
    checksum += int(card[-i-1])

print(checksum)
# - rule 3: If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
if checksum % 10 != 0:
    print("INVALID")
elif length == 15 and int(str(card)[:2]) in [34, 37]:
    print("AMEX")
elif length == 16 and int(str(card)[:2]) in [51, 52, 53, 54, 55]:
    print("MASTERCARD")
elif length in [13, 16] and int(str(card)[0]) == 4:
    print("VISA")
else:
    print("INVALID")
